_posts/2019-12-29-working-with-interfaces-in-go.md

```diff
@@ -0,0 +1,140 @@
+---
+layout: post
+title: Working with Interfaces in Go
+categories: [golang]
+---
+
+[Interfaces](https://golang.org/doc/effective_go.html#interfaces_and_types) are my favorite feature in Go. An interface type represents a set of methods. Unlike most other languages, you don’t have to explicitly declare that a type *implements* an interface. A struct **`S`** implements the interface **`I`** implicitly if **`S`** defines the methods that **`I`** requires. 
+
+Writing good interfaces is difficult. It's easy to [pollute](https://rakyll.org/interface-pollution/) the "API" of a package by exposing broad or unnecessary interfaces. In this article, we'll explain the reasoning behind existing guidelines for interfaces and supplement them with examples from the standard library. We'll also observe other patterns used in the library that don't conform to these best practices.
+
+#### [guideline] [The bigger the interface, the weaker the abstraction](https://www.youtube.com/watch?v=PAAkCSZUG1c&t=5m17s)
+
+It's unlikely that you'll be able to find multiple types that can implement a large interface. For that reason, ”interfaces with only one or two methods are common in Go code”. Instead of declaring large *public* interfaces, consider depending on or returning an explicit type. 
+
+The [`io.Reader` and `io.Writer`](https://github.com/golang/go/blob/c170b14c2c1cfb2fd853a37add92a82fd6eb4318/src/io/io.go#L77-L92) interfaces are the usual examples for powerful interfaces. 
+
+```go
+type Reader interface {
+	Read(p []byte) (n int, err error)
+}
+```
+
+After grepping the std lib, I found 81 structs across 30 packages that implement an `io.Reader`, and 99 methods or functions that consume it across 39 packages.
+
+#### [guideline] [Go interfaces *generally* belong in the package that uses values of the interface type, not the package that implements those values](https://github.com/golang/go/wiki/CodeReviewComments#interfaces)
+
+By defining the interface in the package that actually uses it, we can guarantee that we're not defining an unnecessary interface.
```

Comment:
Consider changing `we can guarantee that we're not defining an unnecessary interface.` to something like `we can let the client define the abstraction rather than the provider dictating the abstraction to all clients.` or something along those lines and even link out to the [tweet](https://twitter.com/davecheney/status/942593128355192832) or quote however you prefer.

Reply:
Done!